enginepb: clean up MVCCValueHeaderPure hack

RaduBerinde opened the pull request:
This change updates `MVCCValueHeader` to use the new
`gogoproto.omitempty` option from our fork to avoid encoding empty
fields that are not nullable. Without this option, even an empty
non-scalar field will result in a couple of encoded bytes (the tag and
the length).

We remove `MVCCValueHeaderPure` and `MVCCValueHeaderCrdbTest` which
were variants used to generate different versions of the marshaling
code.

The encoding should be unchanged. Performance is improved because we
no longer have to make a copy into `MVCCValueHeaderPure`.

```
name                                                                   old time/op    new time/op    delta
EncodeMVCCValue/header=local_walltime+logical/value=short-8              44.4ns ± 0%    35.6ns ± 0%  -19.94%  (p=0.000 n=7+8)
EncodeMVCCValue/header=local_walltime+logical/value=long-8                651ns ± 6%     612ns ± 4%   -5.98%  (p=0.005 n=8+8)
EncodeMVCCValue/header=local_walltime+logical/value=tombstone-8          43.2ns ± 0%    34.2ns ± 0%  -20.95%  (p=0.000 n=7+8)
EncodeMVCCValue/header=omit_in_rangefeeds/value=short-8                  34.1ns ± 3%    29.2ns ± 2%  -14.56%  (p=0.001 n=7+7)
EncodeMVCCValue/header=omit_in_rangefeeds/value=long-8                    589ns ± 9%     605ns ± 3%     ~     (p=0.130 n=8+8)
EncodeMVCCValue/header=omit_in_rangefeeds/value=tombstone-8              29.9ns ± 4%    25.7ns ± 0%  -14.18%  (p=0.000 n=8+8)
EncodeMVCCValue/header=local_walltime+jobID/value=tombstone-8            42.3ns ± 1%    34.9ns ± 3%  -17.56%  (p=0.000 n=7+8)
EncodeMVCCValue/header=local_walltime+jobID/value=short-8                43.1ns ± 0%    37.0ns ± 6%  -14.08%  (p=0.000 n=8+8)
EncodeMVCCValue/header=local_walltime+jobID/value=long-8                  628ns ± 2%     640ns ± 7%     ~     (p=0.161 n=8+8)
EncodeMVCCValue/header=local_walltime+logical+jobID/value=tombstone-8    44.4ns ± 0%    36.4ns ± 4%  -18.11%  (p=0.000 n=8+8)
EncodeMVCCValue/header=local_walltime+logical+jobID/value=short-8        45.6ns ± 2%    36.6ns ± 1%  -19.70%  (p=0.001 n=7+7)
EncodeMVCCValue/header=local_walltime+logical+jobID/value=long-8          600ns ± 4%     634ns ±15%     ~     (p=0.072 n=7+8)
EncodeMVCCValue/header=empty/value=short-8                               8.15ns ± 1%    8.24ns ± 6%     ~     (p=0.488 n=8+8)
EncodeMVCCValue/header=empty/value=long-8                                8.02ns ± 2%    8.15ns ± 7%     ~     (p=0.721 n=8+8)
EncodeMVCCValue/header=empty/value=tombstone-8                           7.84ns ± 0%    8.10ns ± 1%   +3.30%  (p=0.001 n=7+7)
EncodeMVCCValue/header=jobID/value=tombstone-8                           30.8ns ± 0%    26.9ns ± 2%  -12.45%  (p=0.000 n=8+8)
EncodeMVCCValue/header=jobID/value=short-8                               36.3ns ± 0%    30.1ns ± 0%  -17.05%  (p=0.000 n=8+8)
EncodeMVCCValue/header=jobID/value=long-8                                 602ns ± 1%     577ns ± 3%   -4.12%  (p=0.001 n=6+8)
EncodeMVCCValue/header=local_walltime/value=tombstone-8                  41.7ns ± 1%    32.4ns ± 0%  -22.49%  (p=0.000 n=8+7)
EncodeMVCCValue/header=local_walltime/value=short-8                      42.2ns ± 0%    33.5ns ± 0%  -20.77%  (p=0.000 n=7+8)
EncodeMVCCValue/header=local_walltime/value=long-8                        640ns ± 7%     643ns ± 1%     ~     (p=0.718 n=8+7)
EncodeMVCCValueForExport/header=empty/value=tombstone-8                  6.61ns ± 3%    6.55ns ± 1%     ~     (p=0.646 n=7+7)
EncodeMVCCValueForExport/header=empty/value=short-8                      6.52ns ± 1%    6.51ns ± 1%     ~     (p=0.196 n=7+7)
```

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/135628)
<!-- Reviewable:end -->

RaduBerinde committed:
enginepb: clean up MVCCValueHeaderPure hack

This change updates `MVCCValueHeader` to use the new
`gogoproto.omitempty` option from our fork to avoid encoding empty
fields that are not nullable. Without this option, even an empty
non-scalar field will result in a couple of encoded bytes (the tag and
the length).

We remove `MVCCValueHeaderPure` and `MVCCValueHeaderCrdbTest` which
were variants used to generate different versions of the marshaling
code.

The encoding should be unchanged. Performance is improved because we
no longer have to make a copy into `MVCCValueHeaderPure`.

```
name                                                                   old time/op    new time/op    delta
EncodeMVCCValue/header=local_walltime+logical/value=short-8              44.4ns ± 0%    35.6ns ± 0%  -19.94%  (p=0.000 n=7+8)
EncodeMVCCValue/header=local_walltime+logical/value=long-8                651ns ± 6%     612ns ± 4%   -5.98%  (p=0.005 n=8+8)
EncodeMVCCValue/header=local_walltime+logical/value=tombstone-8          43.2ns ± 0%    34.2ns ± 0%  -20.95%  (p=0.000 n=7+8)
EncodeMVCCValue/header=omit_in_rangefeeds/value=short-8                  34.1ns ± 3%    29.2ns ± 2%  -14.56%  (p=0.001 n=7+7)
EncodeMVCCValue/header=omit_in_rangefeeds/value=long-8                    589ns ± 9%     605ns ± 3%     ~     (p=0.130 n=8+8)
EncodeMVCCValue/header=omit_in_rangefeeds/value=tombstone-8              29.9ns ± 4%    25.7ns ± 0%  -14.18%  (p=0.000 n=8+8)
EncodeMVCCValue/header=local_walltime+jobID/value=tombstone-8            42.3ns ± 1%    34.9ns ± 3%  -17.56%  (p=0.000 n=7+8)
EncodeMVCCValue/header=local_walltime+jobID/value=short-8                43.1ns ± 0%    37.0ns ± 6%  -14.08%  (p=0.000 n=8+8)
EncodeMVCCValue/header=local_walltime+jobID/value=long-8                  628ns ± 2%     640ns ± 7%     ~     (p=0.161 n=8+8)
EncodeMVCCValue/header=local_walltime+logical+jobID/value=tombstone-8    44.4ns ± 0%    36.4ns ± 4%  -18.11%  (p=0.000 n=8+8)
EncodeMVCCValue/header=local_walltime+logical+jobID/value=short-8        45.6ns ± 2%    36.6ns ± 1%  -19.70%  (p=0.001 n=7+7)
EncodeMVCCValue/header=local_walltime+logical+jobID/value=long-8          600ns ± 4%     634ns ±15%     ~     (p=0.072 n=7+8)
EncodeMVCCValue/header=empty/value=short-8                               8.15ns ± 1%    8.24ns ± 6%     ~     (p=0.488 n=8+8)
EncodeMVCCValue/header=empty/value=long-8                                8.02ns ± 2%    8.15ns ± 7%     ~     (p=0.721 n=8+8)
EncodeMVCCValue/header=empty/value=tombstone-8                           7.84ns ± 0%    8.10ns ± 1%   +3.30%  (p=0.001 n=7+7)
EncodeMVCCValue/header=jobID/value=tombstone-8                           30.8ns ± 0%    26.9ns ± 2%  -12.45%  (p=0.000 n=8+8)
EncodeMVCCValue/header=jobID/value=short-8                               36.3ns ± 0%    30.1ns ± 0%  -17.05%  (p=0.000 n=8+8)
EncodeMVCCValue/header=jobID/value=long-8                                 602ns ± 1%     577ns ± 3%   -4.12%  (p=0.001 n=6+8)
EncodeMVCCValue/header=local_walltime/value=tombstone-8                  41.7ns ± 1%    32.4ns ± 0%  -22.49%  (p=0.000 n=8+7)
EncodeMVCCValue/header=local_walltime/value=short-8                      42.2ns ± 0%    33.5ns ± 0%  -20.77%  (p=0.000 n=7+8)
EncodeMVCCValue/header=local_walltime/value=long-8                        640ns ± 7%     643ns ± 1%     ~     (p=0.718 n=8+7)
EncodeMVCCValueForExport/header=empty/value=tombstone-8                  6.61ns ± 3%    6.55ns ± 1%     ~     (p=0.646 n=7+7)
EncodeMVCCValueForExport/header=empty/value=short-8                      6.52ns ± 1%    6.51ns ± 1%     ~     (p=0.196 n=7+7)
```

Epic: none
Release note: None

tbg reviewed (approved):
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135628)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @sumeerbhola)

___
*[`pkg/storage/enginepb/mvcc3.proto` line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135628#-OC3xK11BMC-vL0U3YMP:-OC3xK11BMC-vL0U3YMQ:b-tx80ud) ([raw file](https://github.com/cockroachdb/cockroach/blob/128db0a93fffa3f84f49c1189a8c1da03560e19f/pkg/storage/enginepb/mvcc3.proto#L221)):*
> ```proto
> }
> 
> // MVCCValueHeaderPure is not to be used directly. It's generated only for use of
> ```

:heart: I love to see this disappear.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

bors r+
